src/compiler/checker.ts

```diff
@@ -19350,6 +19360,14 @@ namespace ts {
                         if (isMatchingReference(reference, right)) {
                             return narrowTypeByEquality(type, operator, left, assumeTrue);
                         }
+                        if (assumeTrue && (operator === SyntaxKind.EqualsEqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsEqualsToken)) {

```

Comment:
Optional Chaining only affects types if `strictNullChecks` is enabled. As an optimization, would it make sense to skip this check when not in `strictNullChecks`?